File: go/11.container-with-most-water.go
/*
 * @lc app=leetcode id=11 lang=golang
 *
 * [11] Container With Most Water
 */
 func maxArea(height []int) int {
	water := 0
	i := 0
	j := len(height) - 1

	for i < j {
		water = max(water, min(height[i], height[j]) * (j - i))
		if height[i] < height[j] {
			i += 1
		} else {
			j -= 1
		}
	}

	return water
}

func max(a, b int) int {
	if a > b {
		return a
	} else {
		return b
	}
}

func min(a, b int) int {
	if a < b {
		return a
	} else {
		return b
	}
}

File: rust/11.container-with-most-water.rs
/*
 * @lc app=leetcode id=11 lang=rust
 *
 * [11] Container With Most Water
 */
impl Solution {
    pub fn max_area(height: Vec<i32>) -> i32 {
        let mut i: usize = 0;
        let mut j: usize = (height.len() - 1) as usize;

        let mut water = 0;
        while i < j {
            water = water.max((j - i) * height[i].min(height[j]) as usize);
            if height[i] < height[j] {
                i = i + 1;
            } else {
                j = j - 1;
            }
        }
        water as i32
    }
}

